Fix occupied resource releasing if app already unregistered

AKalinich-Luxoft opened the pull request:
Fixed occupied resource releasing in case its owning app currently not present in `ApplicationManager`. The problem was that if some rc application acquires some of resources using `Set/ButtonPress` and then disconnects - the second rc app could not be able to acquire same resource at all.

Following changes were done:
- Added missing description for some functions
- Updated virtual destructors definition in `ResourceAllocationManager` class hierarhy
- Added `OnUnregisterApplication()` function to `PluginManager` to notify its plugins about app unregistration
- Added `OnUnregisterApplication()` function to `ResourceAllocationManager` to automatically release
allocated resources for application which is unregistered
- Fixed UT, mock, added new UT for verifying defect

AKalinich-Luxoft commented:
@LuxoftAKutsan, @dev-gh, @LitvinenkoIra please review

dev-gh commented on src/components/remote_control/src/resource_allocation_manager_impl.cc:
```diff
@@ -200,5 +216,27 @@ void ResourceAllocationManagerImpl::OnDriverDisallowed(
 
-ResourceAllocationManager::~ResourceAllocationManager() {}
+void ResourceAllocationManagerImpl::RefreshAllocatedResourcesState() {
+  LOG4CXX_AUTO_TRACE(logger_);
```
@AKalinich-Luxoft 
Suggest to investigate approach when AM notifies plugin on application unregistration. There is a similar implementation for OnDeviceRemoved notification. Otherwise you'll have to not forget to do manually that cleanup in every other place/class.

AKalinich-Luxoft replied:
@dev-gh accepted. Now added notification AM->PluginManager->RC plugin->ResourceAllocationManager when application was unregistered so there is no more need to do related cleanup manually in all places. Please see https://github.com/LuxoftAKutsan/sdl_core/pull/199/commits/a4d880829583b394f673410fb577272cc1ebf3fa

AKalinich-Luxoft committed:
Fix occupied resource releasing if app already unregistered

Fixed occupied resource releasing in case its owning app
currently not present in ApplicationManager. The problem
was that if some rc application acquires some of resources
using Set/ButtonPress and then disconnects - the second rc
app could not be able to acquire same resource at all.

Following changes were done:
- Added missing description for some functions
- Updated virtual destructors definition in
ResourceAllocationManager class hierarhy
- Added OnUnregisterApplication() function to PluginManager
to notify its plugins about app unregistration
- Added OnUnregisterApplication() function to
ResourceAllocationManager to automatically release
allocated resources for application which is unregistered
- Fix rejected response if resource is already free

AKalinich-Luxoft committed:
Fix UT, mocks, added new UT for verifying defect

dev-gh commented on src/components/application_manager/src/application_manager_impl.cc:
```diff
@@ -2768,6 +2768,12 @@ void ApplicationManagerImpl::UnregisterApplication(
   }
 
+#ifdef SDL_REMOTE_CONTROL
+  if (!is_resuming) {
```
@AKalinich-Luxoft 
I don't think you need this check. If application is unregistered no matter of what reason plugin must be notified

AKalinich-Luxoft replied:
@dev-gh fixed in https://github.com/LuxoftAKutsan/sdl_core/pull/199/commits/fcc3f2c1c2e14c061c7aee64e28da99b86fa211d

dev-gh commented on src/components/remote_control/src/resource_allocation_manager_impl.cc:
```diff
@@ -200,5 +203,24 @@ void ResourceAllocationManagerImpl::OnDriverDisallowed(
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  rejected_resources_for_application_.erase(app_id);
+  for (AllocatedResources::const_iterator it = allocated_resources_.begin();
```
@AKalinich-Luxoft 
Seems like std::remove_if can be used here

AKalinich-Luxoft replied:
@dev-gh seems that `std::remove_if` is not suitable for associative containers like `std::map`
More information you can find [here](http://www.cplusplus.com/forum/general/131917/)

AKalinich-Luxoft commented:
Rebuild required
